### PR TITLE
feat(emails): notify users when they are running low on recovery codes

### PR DESCRIPTION
**Pull request by vbudhram**
Connects to https://github.com/mozilla/fxa-auth-server/issues/2409

Blocked on https://github.com/mozilla/fxa-shared/pull/22

Low recovery codes email:
screen shot 2018-05-09 at 9 57 24 pm

### CODE REVIEW

File: test/mailbox.js [hunk 1]
```diff
@@ -45,7 +45,11 @@ module.exports = function (host, port, printLogs) {
         log('mail body', body)
         var json = null
         try {
-          json = JSON.parse(body)[0]
+          json = JSON.parse(body)
+
+          if (json.length === 1) {
```
Comment (vbudhram) [inline test/mailbox.js]: Updated this to return an array of emails (if more than one) or a single email object if only one (backwards compat).

File: test/remote/recovery_email_resend_code_tests.js [hunk 2]
```diff
@@ -24,62 +24,42 @@ describe('remote recovery email resend code', function () {
...
-          }
-        )
-        .then(
-          function (status) {
-            assert.equal(status.verified, true, 'account is verified')
-            assert.equal(status.emailVerified, true, 'account email is verified')
-            assert.equal(status.sessionVerified, true, 'account session is verified')
-          }
-        )
+  it('sign-in verification resend email verify code', () => {
```
Comment (vbudhram) [inline test/remote/recovery_email_resend_code_tests.js]: Needed to update these test cases after adding support for multiple emails.

**Comment by vbudhram** [comment]
@mozilla/fxa-devs r?

File: lib/routes/recovery-codes.js [hunk 3]
```diff
@@ -128,7 +128,7 @@ module.exports = (log, db, config, customs, mailer) => {
               if (remaining === 0) {
```
Comment (shane-tomlinson) [inline lib/routes/recovery-codes.js]: nit: I find the name `remaining` a bit vague. `remainingRecoveryCodes` would be more clear.

File: npm-shrinkwrap.json [hunk 4]
```diff
@@ -195,15 +195,6 @@
       "resolved": "https://registry.npmjs.org/ansi-styles/-/ansi-styles-2.2.1.tgz",
       "integrity": "sha1-tDLdM1i2NM914eRmQ2gkBTPB3b4="
     },
-    "anymatch": {
```
Comment (shane-tomlinson) [inline npm-shrinkwrap.json]: Do you expect -7.4k lines?

File: test/remote/recovery_code_tests.js [hunk 5]
```diff
@@ -138,6 +139,41 @@ describe('remote recovery codes', function () {
...
+        })
+        .then((res) => {
+          assert.equal(res.remaining, 6, 'correct remaining codes')
+          return server.mailbox.waitForEmail(email)
+        })
+        .then((emails) => {
+          // The order in which the emails are sent is not guaranteed, test for both possible templates
+          const email1 = emails[0].headers['x-template-name']
+          const email2 = emails[1].headers['x-template-name']
+          assert.equal(true, email1 === 'postConsumeRecoveryCodeEmail' || email1 === 'lowRecoveryCodesEmail', 'correct template sent')
```
Comment (shane-tomlinson) [inline test/remote/recovery_code_tests.js]: nit: This is super pedantic, this test does not assert that two emails are different.

File: docs/api.md [hunk 6]
```diff
@@ -372,7 +372,7 @@ those common validations are defined here.
 #### lib/devices
 
 * `schema`: {
-    * `id`: isA.string.length(32).regex(HEX_STRING)
+    * `id: isA.string.length(32).regex(HEX_STRING)
```
[the remark this answers is not in the document]
Comment (vbudhram) [reply docs/api.md]: Not sure what happened, but running `node ./scripts/write-api-docs.js` removes the ` character, added it back.

File: lib/routes/recovery-codes.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (vbudhram) [reply lib/routes/recovery-codes.js]: This is similar to https://github.com/mozilla/fxa-content-server/issues/6018, except an email isn't sent, we automatically generate new one when you run out and give them too you during the sign-in flow.

File: docs/api.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (rfk) [reply docs/api.md]: I filed https://github.com/mozilla/fxa-auth-server/issues/2434 to follow up

**Pull request merged**